### PR TITLE
Port plugin-registry to `rust-proto-new`

### PR DESCRIPTION
**Pull request by wimax-grapl**
Doing this as I try and add reflection to our grpc

### CODE REVIEW

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs [hunk 1]
```diff
@@ -1 +1,495 @@
+pub use crate::protobufs::graplinc::grapl::api::plugin_registry::v1beta1::{
```
Comment (jgrillo-grapl) [inline src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs:1]: This breaks encapsulation, which we should not do. The point of `rust-proto-new` is to encapsulate all the yucky protobuf compiler artifacts and expose instead our own SerDe types.

**Review by jgrillo-grapl — CHANGES_REQUESTED** [review]
If there's any way we can keep these protobufs encapsulated that would be really great. I know it's hard work to change this after-the-fact, but basically the entire value-add of this library vs the previous one is that it _does not expose_ any of the generated stubs.

You can see how I achieved this for gRPC APIs in `pipeline-ingress`: https://github.com/grapl-security/grapl/blob/main/src/rust/rust-proto-new/src/graplinc/grapl/api/pipeline_ingress/v1beta1.rs#L174

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs (hunk 1 above)
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs:1]: I'm a tad confused, this is already in rust-proto-new

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_server.rs [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    Serving,
+    NotServing,
+    Unknown,
+}
+
+/**
+ * !!!!! IMPORTANT !!!!!
+ * This is almost entirely cargo-culted from PipelineIngressServer.
+ * Lots of opportunities to deduplicate and simplify.
+ */
```
Comment (wimax-grapl) [inline src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_server.rs:222]: important

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_server.rs [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    service_name: &'static str,
+    e_: PhantomData<E>,
+    f_: PhantomData<F>,
+}
+
+impl<T, E, H, F> PluginRegistryServer<T, E, H, F>
+where
+    T: PluginRegistryApi<E> + Send + Sync + 'static,
+    E: Sync + Send + 'static,
+    tonic::Status: From<E>,
```
Comment (wimax-grapl) [inline src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_server.rs:245]: note that for some reason it was really touchy about Into versus From here

**Review by jgrillo-grapl — CHANGES_REQUESTED** [review]
This looks like it's headed in the right direction! I added a few small questions and comments. I think the last big missing piece is tests. These should include:

unit tests in `rust-proto-new`:
- [x] proptest generators for each SerDe type (see e.g. https://github.com/grapl-security/grapl/blob/main/src/rust/rust-proto-new/tests/integration_test.rs#L51)
- [x] round-trip serde tests for each of the protobufs (see e.g. https://github.com/grapl-security/grapl/blob/main/src/rust/rust-proto-new/tests/integration_test.rs#L258)
- [ ] gRPC tests with mocked business logic which exercise the happy and sad paths of each API (see e.g. https://github.com/grapl-security/grapl/blob/main/src/rust/rust-proto-new/tests/integration_test.rs#L279)

in `plugin-registry`:
- [x] "new-style" integration tests (`#![cfg(feature = "new_integration_tests")]`) which exercise the application business logic via the protocol buffer API client (see e.g. https://github.com/grapl-security/grapl/blob/main/src/rust/pipeline-ingress/tests/integration_test.rs)

File: src/rust/plugin-registry/Cargo.toml [hunk 4]
```diff
@@ -28,8 +28,8 @@ sqlx = { version = "0.5", features = [
 tempfile = "3.3.0"
 thiserror = "1.0.30"
 tokio = { version = "1.17", features = ["full"] }
-tonic = "0.6"
-tonic-health = "0.5.0"
+tonic = { version = "0.7", features = ["default", "compression"] }
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/plugin-registry/Cargo.toml:31]: fixed

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        request: native::CreatePluginRequest,
+    ) -> Result<native::CreatePluginResponse, PluginRegistryServiceClientError> {
+        let response = self
+            .proto_client
+            .create_plugin(proto::CreatePluginRequest::from(request))
+            .await?;
+        let response = native::CreatePluginResponse::try_from(response.into_inner())?;
+        Ok(response)
+    }
+    /// retrieve the plugin corresponding to the given plugin_id
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs:63]: i'm surprised the litner didnt' enforce this tbh

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+use proto::plugin_registry_service_client::PluginRegistryServiceClient as PluginRegistryServiceClientProto;
+use tonic::transport::Endpoint;
+
+use crate::{
+    graplinc::grapl::api::plugin_registry::v1beta1 as native,
+    protobufs::graplinc::grapl::api::plugin_registry::v1beta1 as proto,
+    SerDeError,
+};
+
+const ADDRESS_ENV_VAR: &'static str = "GRAPL_PLUGIN_REGISTRY_ADDRESS";
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs:15]: done

File: src/rust/plugin-registry/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/plugin-registry/Cargo.toml:31]: ah, just kidding - we actually implement Into<tonic::Status> for the PluginRegistryServiceError

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs [hunk 7]
```diff
@@ -1 +1,436 @@
...
+    /// The type of the plugin
+    pub plugin_type: PluginType,
+    /// The byte representation of the plugin executable
+    pub plugin_binary: Vec<u8>,
+}
+
+impl TryFrom<proto::Plugin> for Plugin {
+    type Error = SerDeError;
+
+    fn try_from(value: proto::Plugin) -> Result<Self, Self::Error> {
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs:70]: do you mind terribly if I park that one for when i have nothing else of value to do :)

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+pub struct PluginRegistryServiceClient {
+    proto_client: PluginRegistryServiceClientProto<tonic::transport::Channel>,
+}
+
+impl PluginRegistryServiceClient {
+    /// Create a client from environment
+    #[tracing::instrument(err)]
+    pub async fn from_env() -> Result<Self, Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_client.rs:32]: disagree - intent was to allow configuration in a standard way over env variables - but i dont care enough so doing it

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_server.rs [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+{
+    async fn create_plugin(&self, request: CreatePluginRequest) -> Result<CreatePluginResponse, E>;
+
+    async fn get_plugin(&self, request: GetPluginRequest) -> Result<GetPluginResponse, E>;
+
+    async fn deploy_plugin(&self, request: DeployPluginRequest) -> Result<DeployPluginResponse, E>;
+
+    async fn tear_down_plugin(
+        &self,
+        _request: TearDownPluginRequest,
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1_server.rs:75]: relic from the implementation:
```
    async fn tear_down_plugin(
        &self,
        _request: TearDownPluginRequest,
    ) -> Result<TearDownPluginResponse, PluginRegistryServiceError> {
        todo!()
    }
```

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (colin-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs:70]: I'd honestly rather we keep it as `value` lol it makes it way easier to do batch replacing/ multiple cursor magic

[the remark this answers is not in the document]
Comment (colin-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs:70]: either way tho, var names aren't part of a public api so it seems like something we can punt on

File: src/rust/plugin-registry/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgrillo-grapl) [reply src/rust/plugin-registry/Cargo.toml:31]: This should be addressed by adding a `SerDe` type `Status` type in rust-proto-new and implementing the `From/Into` contract there. We have a story related to this (in the context of some broader library changes) here: https://app.zenhub.com/workspaces/grapl-6036cbd36bacff000ef314f2/issues/grapl-security/issue-tracker/910

File: src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jgrillo-grapl) [reply src/rust/rust-proto-new/src/graplinc/grapl/api/plugin_registry/v1beta1.rs:70]: The point I'm making is it should be consistent with the rest of the library. Whatever choice we make is fine, so long as it's the same choice everywhere (in this library).

File: src/rust/plugin-registry/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgrillo-grapl) [reply src/rust/plugin-registry/Cargo.toml:31]: Looks like this is being addressed in https://github.com/grapl-security/grapl/pull/1665, fine to punt for now but could we add a `TODO` somewhere?

**Review by jgrillo-grapl — APPROVED** [review]
LGTM

File: src/rust/rust-proto-new/src/serde_impl.rs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+};
+use prost::Message;
+
+use crate::{
+    type_url,
+    SerDe,
+    SerDeError,
+};
+
+pub trait HasAssociatedProto<T> {
```
Comment (jgrillo-grapl) [inline src/rust/rust-proto-new/src/serde_impl.rs:13]: Maybe this naming would be more clear?

```rust
pub(crate) trait Serializable<T> {
    type Serialized: From<T> + Message + Default + TryInto<T, Error = SerDeError>;
}
```

File: src/rust/rust-proto-new/src/serde_impl.rs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    type Proto: From<T> + Message + Default + TryInto<T, Error = SerDeError>;
```
Comment (jgrillo-grapl) [inline src/rust/rust-proto-new/src/serde_impl.rs:14]: Why is `Default` here?

Comment (jgrillo-grapl) [inline src/rust/rust-proto-new/src/serde_impl.rs:14]: Definitely want to make sure this doesn't export `prost::Message` from this crate. If `pub(crate)` ensures this then 👍

File: src/rust/rust-proto-new/src/serde_impl.rs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+}
+impl<T> SerDe for T
```
Comment (jgrillo-grapl) [inline src/rust/rust-proto-new/src/serde_impl.rs:16]: This is super cool

**Review by jgrillo-grapl — COMMENTED** [review]
Translating some comments I previously made on a specific commits to an actual code review, because github's UX is very confusing for a comment made on a specific commit...

File: src/rust/rust-proto-new/src/serde_impl.rs (hunk 10 above)
Comment (jgrillo-grapl) [reply src/rust/rust-proto-new/src/serde_impl.rs:13]: Hmm.. I wonder if we can also incorporate the `TypeUrl` thing in here as well like this?

```rust
pub(crate) trait Serializable<T> {
    type Serialized: From<T> + Message + Default + TryInto<T, Error = SerDeError>;

    const TYPE_URL: &'static str;
}
```

File: src/rust/rust-proto-new/src/serde_impl.rs (hunk 11 above)
Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/serde_impl.rs:14]: It's needed for Message::decode, which kinda makes sense (proto deserialization usually constructs an empty/default message and then mutates it)

Comment (wimax-grapl) [reply src/rust/rust-proto-new/src/serde_impl.rs:14]: done (afaik)

**Pull request merged**